core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -460,8 +471,8 @@ private void validateNoNewDeletesForDataFiles(
    * @param dataFilter an expression used to find new conflicting delete files
    */
   protected void validateNoNewDeleteFiles(
-      TableMetadata base, Long startingSnapshotId, Expression dataFilter) {
-    DeleteFileIndex deletes = addedDeleteFiles(base, startingSnapshotId, dataFilter, null);
+      TableMetadata base, Long startingSnapshotId, Expression dataFilter, Snapshot parent) {
```

Comment:
I think a lot of the `revapi` problems are probably related to adding `parent` to these methods. We should add the older version back that passes `base.currentSnapshot()` and mark it deprecated, to be removed in 1.3.0. That will take care of at least some of the breaking changes.